examples/test_messaging/test_messaging.c

```diff
@@ -0,0 +1,364 @@
+        msg_params->tests_passed = 0;
+        msg_params->tests_failed = 0;
+        msg_params->test_phase = 1;
+        msg_params->address = nf->service_id;
+        msg_params->msg_pool = nf_msg_pool;
+        msg_params->msg_q = nf->msg_q;
+        msg_params->ring_count = 0;
+        msg_params->mempool_count = 0;
+        msg_params->test_msg_count = 0;
+        nf->data = (void *)msg_params;
+
+        while ((c = getopt(argc, argv, "p:")) != -1) {
+                switch (c) {
+                        case 'p':
+                                print_delay = strtoul(optarg, NULL, 10);
+                                RTE_LOG(INFO, APP, "print_delay = %d\n", print_delay);
+                                break;
+                        case '?':
+                                usage(progname);
+                                if (optopt == 'p')
+                                        RTE_LOG(INFO, APP, "Option -%c requires an argument.\n", optopt);
+                                else if (isprint(optopt))
+                                        RTE_LOG(INFO, APP, "Unknown option `-%c'.\n", optopt);
+                                else
+                                        RTE_LOG(INFO, APP, "Unknown option character `\\x%x'.\n", optopt);
+                                return -1;
+                        default:
+                                usage(progname);
+                                return -1;
+                }
+        }
+        return optind;
+}
+
+/*
+ * Sets up the NF. Initiates Unit Test
+ */
+void
+nf_setup(struct onvm_nf_local_ctx *nf_local_ctx) {
+        struct rte_mempool *pktmbuf_pool;
+        pktmbuf_pool = rte_mempool_lookup(PKTMBUF_POOL_NAME);
+        if (pktmbuf_pool == NULL) {
+                onvm_nflib_stop(nf_local_ctx);
+                rte_exit(EXIT_FAILURE, "Cannot find mbuf pool!\n");
+        }
+
+        printf("\nTEST MESSAGING STARTED\n");
+        printf("---------------------------\n");
+        struct test_msg_data *msg_params;
+        msg_params = (struct test_msg_data *)nf_local_ctx->nf->data;
+        msg_params->mempool_count = rte_mempool_avail_count(msg_params->msg_pool);
+        int *msg_int;
+        msg_int = (int*)rte_malloc(NULL, sizeof(int), 0);
+        if (msg_int == NULL) {
+                printf("Message was not able to be malloc'd\n");
+        }
+        *msg_int = MAGIC_NUMBER;
+        int ret = onvm_nflib_send_msg_to_nf(msg_params->address, msg_int);
+        if (ret != 0) {
+                printf("Message was unable to be sent\n");
+        }
+}
+
+/*
+ * Runs logic for tests
+ */
+void
+nf_msg_handler(void *msg_data, struct onvm_nf_local_ctx *nf_local_ctx) {
+        struct test_msg_data *msg_params;
+        msg_params = (struct test_msg_data *)nf_local_ctx->nf->data;
+        // Tests if one message can be sent to itself
+        if (1 == msg_params->test_phase) {
```

Comment:
I think that it would be helpful from a readability standpoint to break out the test phases into their own functions, i.e test_send_receive_single_message(msg_params). Non-blocking to getting this merged.